### DISCUSSION TITLE
Database.forConfig doesn't support transaction.

### DISCUSSION THREAD

**Discussion by jeffwji** [slick/slick#2414] [General]
I have two test cases for transaction testing:

The first test case:

```scala
    implicit val dc: DatabaseConfig[JdbcProfile] = DatabaseConfig.forConfig[JdbcProfile]("database.config")
    import dc.profile.api._

    // get db from URL
    val db: dc.profile.backend.DatabaseDef = Database.forURL(
      "jdbc:h2:mem:;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE;IGNORECASE=TRUE;DATABASE_TO_UPPER=false;INIT=RUNSCRIPT FROM 'classpath:database/h2_init.sql'"
      , driver="org.h2.Driver"
    )

    val cityDao = CityDao
    val id = UUID.randomUUID().toString

    val creation = cityDao.create(new City(id, "NotExistingCity"))
        .flatMap { _ => DBIO.failed(new Exception("Roll it back")) }

    Await.result(db.run(creation.transactionally), Duration.Inf)
    Await.result(db.run(cityDao.selectById(UUID.fromString(id)).map{ records =>
      assert(records.isEmpty)
    }), Duration.Inf)
```

This one perfectly works, but the following one, the only difference is the `db` is getting from `Config` instead of from URL, it doesn't work properly:

```scala
    implicit val dc: DatabaseConfig[JdbcProfile] = DatabaseConfig.forConfig[JdbcProfile]("database.config")
    import dc.profile.api._

    // get db from Config
    val db: dc.profile.backend.DatabaseDef = Database.forConfig("database.config.db")  // Or `val db = dc.db`

    val cityDao = CityDao
    val id = UUID.randomUUID().toString

    val creation = cityDao.create(new City(id, "NotExistingCity"))
        .flatMap { _ => DBIO.failed(new Exception("Roll it back")) }

    Await.result(db.run(creation.transactionally), Duration.Inf)
    Await.result(db.run(cityDao.selectById(UUID.fromString(id)).map{ records =>
      assert(records.isEmpty)
    }), Duration.Inf)
```

I tried to get db directly from `databaseConfig` like this: `val db = dc.db`. It doesn't work either, that's why I switch to get the config directly, but still doesn't work.

The content of configuration has not difference to URL:

```conf
database.config = {
  profile = "slick.jdbc.H2Profile$"
  db = {
    driver = "org.h2.Driver"
    url = "jdbc:h2:mem:;DB_CLOSE_DELAY=-1;DB_CLOSE_ON_EXIT=FALSE;IGNORECASE=TRUE;DATABASE_TO_UPPER=false;INIT=RUNSCRIPT FROM 'classpath:database/h2_init.sql'"
  }
}
```

I'm using Slick version 3.3.3

Database init script:

```sql
CREATE TABLE IF NOT EXISTS city(
  id UUID PRIMARY KEY
  ,name VARCHAR(255) NOT NULL
);
```

**Comment by jeffwji** [comment]
I found the difference between these two ways, `forURL` does not use `HikariCP` as connection pool by default, so if I disable it from config, `forConfig` will work. But if saying `HikariCP` doesn't support transaction, which is still not make sense.

**Comment by nafg** [comment]
I'm a little confused. But `  _ = DBIO.failed(new Exception("Roll it back"))` shouldn't do anything. You should use `<-` instead of `=`.

The first way, the desugaring would be something like

```scala
cityDao.create(new City(id, "NotExistingCity"))
  .map { _affect =>
    val ignore = DBIO.failed(new Exception("Roll it back"))
    ()
  }
```

So it's just a value that the DBIO interpreter never sees.

With `<-` it would be more like

```scala
cityDao.create(new City(id, "NotExistingCity"))
  .flatMap { _affect =>
    DBIO.failed(new Exception("Roll it back"))
  }
  .map(_ => ())
```

So the DBIO interpreter knows to go to failed state.

**Reply by jeffwji** [reply]
Yes, I saw that issue, and the post has been updated

**Comment by nafg** [comment]
What exactly is `assertion: Assertion = assert(records.isEmpty)`? Is this scalatest's `Assertion` type and `assert` method?

More to the point, is assert throwing an exception or just returning a value?

Anyway if you're still stuck, can you put compiling code on Scastie for me to see?

**Reply by jeffwji** [reply]
Yes, it is a test case

**Reply by jeffwji** [reply]
It was failed with:

```
Vector(CreateBusinessAccountEvent.City(id=445363e9-53bb-4a8d-b9e1-3500a6ca99f0, name=NotExistingCity)) was not empty
ScalaTestFailureLocation: repository.dao.CityDaoTestSuite at (CityDaoTestSuite.scala:52)
org.scalatest.exceptions.TestFailedException: Vector(CreateBusinessAccountEvent.City(id=445363e9-53bb-4a8d-b9e1-3500a6ca99f0, name=NotExistingCity)) was not empty
	at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
	at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
	at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
	at repository.dao.CityDaoTestSuite.$anonfun$new$9(CityDaoTestSuite.scala:52)
	at slick.dbio.DBIOAction.$anonfun$map$1(DBIOAction.scala:43)
	at slick.basic.BasicBackend$DatabaseDef.$anonfun$runInContextInline$1(BasicBackend.scala:172)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:434)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

**Reply by jeffwji** [reply]
This is actually complete code already:

```scala
class CityDaoTestSuite extends AnyWordSpecLike with StrictLogging{
  import scala.concurrent.ExecutionContext.Implicits.global

  "create" should {
    "rollback if exception happens" in {
      implicit val dc: DatabaseConfig[JdbcProfile] = DatabaseConfig.forConfig[JdbcProfile]("database.config")
      import dc.profile.api._

      val db = dc.db
      val cityDao = CityDao
      val id = UUID.randomUUID().toString

      val creation = cityDao.create(new City(id, "NotExistingCity"))
        .flatMap { _ => DBIO.failed(new Exception("Roll it back")) }

      Await.result(db.run(creation.transactionally), Duration.Inf)
      Await.result(db.run(cityDao.selectById(UUID.fromString(id)).map{ records =>
        assert(records.isEmpty)
      }), Duration.Inf)
    }
}
```

**Comment by jeffwji** [comment]
All right, got some progress, the `=` does do nothing to rollback, that's why insertion works all the time, after change to `flatMap`, it throws 

```
Roll it back
java.lang.Exception: Roll it back
	at repository.dao.CityDaoTestSuite.$anonfun$new$12(CityDaoTestSuite.scala:69)
	at repository.dao.CityDaoTestSuite.$anonfun$new$12$adapted(CityDaoTestSuite.scala:69)
```

What if I just want it to be failure(rollback) without exception?

**Reply by jeffwji** [reply]
Now, I have to use `asTry` and ignore the failure, is this the only way that I can skip the exception?

```scala
implicit val dc: DatabaseConfig[JdbcProfile] = DatabaseConfig.forConfig[JdbcProfile]("database.config")
import dc.profile.api._

val db = dc.db
val cityDao = CityDao
val id = UUID.randomUUID().toString

val creation = cityDao.create(new City(id, "NotExistingCity"))
  .flatMap { _ => DBIO.failed(new Exception("Roll it back")) }

Await.result(db.run{
  creation.transactionally.asTry.flatMap{ _: Try[Nothing] =>
    cityDao.selectById(UUID.fromString(id)).map{ records =>
      assert(records.isEmpty)
    }
  }
}, Duration.Inf)
```

**Reply by nafg** [reply]
> What if I just want it to be failure(rollback) without exception?

Not sure I understand the question

**Reply by nafg** [reply]
> Now, I have to use `asTry` and ignore the failure, is this the only way that I can skip the exception?

DBIO has a bunch of combinators that allow you to deal with failures in different ways. I'm not that expert in all of them. And it certainly lacks some that Future and other effect types have.

Is there something specific that you want to improve about your code, or are you just asking "what are all the ways"?

**Reply by jeffwji** [reply]
I mean is throwing an exception the only way to get the program to rollback?

**Reply by nafg** [reply]
It's not throwing an exception to create one and pass it to DBIO.failed. Returning DBIO.failed is the only way I know of. I see your point though that it makes continuing more complicated.

If you have an idea for a better API, a PR is welcome.

**Reply by jeffwji** [reply]
DBIO.failed will raise the exception, which stop the following process, that's why I have to wrap it with Try. But it look's ugly. I'm hoping there's any other way to rollback the transaction without create exception.

**Reply by nafg** [reply]
One option would be to define a singleton object `Rollback` that extends `scala.util.control.NoStackTrace` or calls the Throwable constructor with `writableStackTrace = false`. Then there could be a `DBIO.rollback` method that is a shorthand for `DBIO.failed(Rollback)`.

Either way, you should probably never ignore other exceptions. And if the success case produces a value, rollback wouldn't have it because it would be a different path, so you need to deal with the different "what ifs."

In terms of your code, a shorthand for `.flatMap(_ => ...)` is `>>` so you could change

```scala
 creation.transactionally.asTry.flatMap{ _ =>
    cityDao.selectById(UUID.fromString(id)).map{ records =>
      assert(records.isEmpty)
    }
  }
```

to

```scala

 creation.transactionally.asTry >>
    cityDao.selectById(UUID.fromString(id)).map{ records =>
      assert(records.isEmpty)
    }
```

**Reply by jeffwji** [reply]
`NoStackTrace` just an special `Throwable` without track trace, it does not prevent throwing. in this case `asTry` is still needed.

**Reply by jeffwji** [reply]
Rollback operation could be an intentionally action, not have to be exception or error.

**Reply by nafg** [reply]
`asTry` would still be needed, but being able to write `DBIO.rollback` instead of `DBIO.failed(new Exception(...))` might be an improvement

Having standard error handling operations like Future and IO, such `recover`, `recoverWith`, etc. would also be nice.

But if you want to suggest a fundamentally different model for implementing rollback in the context of DBIO I would need to hear a more concrete proposal in order to have an opinion on it.

**Reply by jeffwji** [reply]
Do I need import any thing to enable `DBIO.rollback`? I can't see this function from DBIO even if I defined `Rollback` in the context.

**Comment by nafg** [comment]
It doesn't exist, it was suggestion for something to add to slick

On Fri Mar 25, 2022, 02:36 AM GMT, jeffwji ***@***.***> wrote:

Do I need import any thing to enable DBIO.rollback? I can't see this
function from DBIO even if I defined Rollback in the context.

—
Reply to this email directly, view it on GitHub
<https://github.com/slick/slick/discussions/2414#discussioncomment-2434599>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AAAYAUDYYWY7M4QRCOGGUJ3VBURERANCNFSM5RPPK7FA>
.
You are receiving this because you commented.Message ID:
***@***.***>
